src/akplan/solve.py

```diff
@@ -236,62 +236,58 @@ def _init_lower_upper(coords: list[pd.Index]) -> tuple[xr.DataArray, xr.DataArra
+                time.loc[ak_id, timeslot_id] + time.loc[ak_id, other_timeslots] <= 1,
+                name=_construct_constraint_name(
+                    "AKContiguous",
+                    ak_id,
+                    block_id,
+                    timeslot_id,
+                ),
+            )
     logger.debug("Constraints AKContiguous added")
 
-    # TODO vectorize
-    if input_data.config.max_num_timeslots_before_break > 0:
+    # shorter alias
+    num_max_consec = input_data.config.max_num_timeslots_before_break
+    if num_max_consec > 0:
         # PersonNeedsBreak
         # Any real person needs a break after some number of time slots
         # So in each block at most  consecutive timeslots can be active for any person
-        for block_entry in ids.block_dict.values():
-            for idx in range(
-                len(block_entry) - input_data.config.max_num_timeslots_before_break - 1
-            ):
-                block_subset = block_entry[
-                    idx : idx + input_data.config.max_num_timeslots_before_break + 1
-                ]
-                m.add_constraints(
-                    lhs=person_time.loc[:, block_subset].sum("timeslot"),
-                    sign="<=",
-                    rhs=input_data.config.max_num_timeslots_before_break,
-                    name=_construct_constraint_name("BreakForPerson", block_entry[idx]),
-                )
+
+        # expand person_time to additional block idx
+        block_person_time = props.block_mask * person_time
+
+        # we assume that timeslots are chronologically consecutive
+        rolling_sum = block_person_time.rolling(timeslot=num_max_consec + 1).sum()
+
+        # ignore partial rolling at the beginning of blocks
+        ignore_fist_n_mask = (
+            props.block_mask.cumsum(dim="timeslot") > num_max_consec
+        )
+        m.add_constraints(
+            rolling_sum.where(props.block_mask & ignore_fist_n_mask) <= num_max_consec,
+            name="BreakForPerson",
+        )
     logger.debug("Constraints BreakForPerson added")
 
-    # TODO vectorize
     # AK dependencies
+
+    # calculate cumulative sum. We use `isel` to index 'from the left' of window
+    time_tail_sum = time.isel(timeslot=slice(None, None, -1)).cumsum("timeslot")
     for ak_id in ids.ak:
         if ak_id not in props.dependencies:
             continue
-        other_ak_ids = props.dependencies[ak_id]
-        for idx, timeslot_id in enumerate(ids.timeslot):
-            m.add_constraints(
-                lhs=time.loc[ak_id, ids.timeslot[idx:]].sum("timeslot")
-                - time.loc[other_ak_ids, timeslot_id],
-                sign=">=",
-                rhs=0,
-                name=_construct_constraint_name(
-                    "AKDependenciesDoneBeforeAK", ak_id, timeslot_id
-                ),
-            )
+        m.add_constraints(
+            time_tail_sum.loc[ak_id] - time.loc[props.dependencies[ak_id]] >= 0,
```

Comment:
```suggestion
            len(props.dependencies[ak_id])*time_tail_sum.loc[ak_id] - time.loc[props.dependencies[ak_id]] >= 0,
```
If there are more dependencies than the `ak_durations[ak_id]` we need this addional factor. Or we split this inqeuality into one inequality per dependency as done in the wiki.

Reply:
I think if there are a lot of dependencies, splitting the inequality is the better option to avoid big-M constraints.

Reply:
Ah, we might want to add an explanatory comment. For your interpretation to be correct, we would need to write (note the `sum` at the end)
```python
time_tail_sum.loc[ak_id] - time.loc[props.dependencies[ak_id]].sum("timeslot") >= 0
```

As it stands the term `time_tail_sum.loc[ak_id]` has only the coord `timeslot`; the term `time.loc[...]` has coords `ak` and `timeslot`. So broadcasting yields one constraint per dependency and timeslot

Reply:
Then I think we can leave the code as is, add a comment and change the wiki accordingly.